### PR TITLE
Duplicate song warning added 

### LINKED ISSUE #554: Feature request: Warn when adding song that already exists in music playlist
When adding a song to a music playlist, it would be very helpful to receive a warning if that song is already present in the playlist, with the ability to abort the operation and not add a duplicate song to the playlist.

A use case for this feature is as follows: I own many albums and have curated a playlist of the songs I want in rotation while driving. This playlist has around 1000 songs. When I am playing a whole album and I hear a song that I want to be present in that playlist, I try to add it to the playlist. If it is not in the playlist, it is added. If it is already in the playlist, I see a warning and can choose not to add a duplicate.

For example, the now-defunct Google Play Music app issued this warning when the user attempted to add a song which was already present in the playlist:

duplicate_song_warning

Of course, the requested warning could only be seen when using the Navidrome UI and not a third-party Subsonic client like DSub or Audinaut.

### PR DESCRIPTION
**Pull request by RitikPandey1**
Fixes #554 

If user skip :
navidromeSkip

If user want to add song : 
navidromeAdd

Tell me if any changes are required @deluan 
Happy to contribute :)

### CODE REVIEW

**Review by deluan — CHANGES_REQUESTED** [review]
Hey @RitikPandey1 , this looks great! A couple of notes, though:
- I think it would be better to leave the "Add" button of the `AddToPlaylist` dialog disabled, and only enable it if the `checkDuplicateSong` determine it is not duplicated, or if the user clicks "Add" in the DuplicateSongDialog
- Ideally we would want to validate if the song is already in the playlist on the server side, but as we are using this issue as a "simple" task to evaluate students for GSoC, a full client-side solution is ok-ish (except when the playlist has thousands of songs, and we will have to retrieve the whole playlist just to check for duplicates.... But this can be tackled later). If you want to try implementing the server side, let me know
- We should extend the check for all songs being added to the playlist, instead of making it work when adding just one song. Maybe this is more complicated for you to tackle. Let me know.

Besides that, the build was not successful, I think because you didn't run `prettier` before commiting. Try configuring your editor to auto run `prettier` on save

Anyway, good job!

**Comment by deluan** [comment]
Regarding my first point in the review, you could do something like what Apple Music does: If at least one music is duplicated, it shows the following dialog:
Screen Shot 2021-03-13 at 9 58 50 PM

The user then can Skip the duplicated, go ahead and Add everything, or just Cancel the whole operation. It is the same dialog regardless if the user is adding one or multiple songs. That simplifies the implementation, and provides a good, uniform UX (IMHO)

**Comment by RitikPandey1** [comment]
> Hey @RitikPandey1 , this looks great! A couple of notes, though:
> 
> * I think it would be better to leave the "Add" button of the `AddToPlaylist` dialog disabled, and only enable it if the `checkDuplicateSong` determine it is not duplicated, or if the user clicks "Add" in the DuplicateSongDialog
> * Ideally we would want to validate if the song is already in the playlist on the server side, but as we are using this issue as a "simple" task to evaluate students for GSoC, a full client-side solution is ok-ish (except when the playlist has thousands of songs, and we will have to retrieve the whole playlist just to check for duplicates.... But this can be tackled later). If you want to try implementing the server side, let me know
> * We should extend the check for all songs being added to the playlist, instead of making it work when adding just one song. Maybe this is more complicated for you to tackle. Let me know.
> 
> Besides that, the build was not successful, I think because you didn't run `prettier` before commiting. Try configuring your editor to auto run `prettier` on save
> 
> Anyway, good job!

Thank you @deluan for review.
Regarding your second point, I'am totally agree with you. Performing duplicate check on client-side is not efficient, but currently I'am learning `go` and not much familiar with it, Once I'll become familiar to `go` then I would love to implement all these features on backend also. I would also like to contribute to navidrome under GSOC.
I have tried my best to  implemented the features(on client-side)  you suggested in first and third point of your review.

**Comment by RitikPandey1** [comment]
> Regarding my first point in the review, you could do something like what Apple Music does: If at least one music is duplicated, it shows the following dialog:
> Screen Shot 2021-03-13 at 9 58 50 PM
> 
> The user then can Skip the duplicated, go ahead and Add everything, or just Cancel the whole operation. It is the same dialog regardless if the user is adding one or multiple songs. That simplifies the implementation, and provides a good, uniform UX (IMHO)

I have added same dialog for single song and multiple songs, this make implementation easy.
Thank you

**Comment by RitikPandey1** [comment]
If user skip duplicate songs then only distinct songs are added in playlist 
navidromeSkip

If user want to add duplicate song : 
navidromeAdd

File: ui/src/i18n/en.json [hunk 1]
```diff
@@ -1,341 +1,346 @@
...
+        "path": "Import from"
+      },
+      "actions": {
+        "selectPlaylist": "Select a playlist:",
+        "addNewPlaylist": "Create \"%{name}\"",
+        "export": "Export"
+      },
+      "message": {
+        "duplicate_song": "Add duplicate song ",
+        "song_exist": "There are duplicates being added to the playlist. Would you like to add the duplicates or skip them ?"
```
Comment (deluan) [inline ui/src/i18n/en.json:129]: Small nitpick: Can you please remove the space before the question mark?

**Review by deluan — COMMENTED** [review]
This looks great. Besides the small nitpick below, I think we can improve the "Skip" behaviour: Instead of returning to the `AddToPlaylist` dialog when clicking the "Skip" button, it should just go ahead and add the non-duplicated songs to the playlist, closing all dialogs, similar to when you press the "Add" button.

File: ui/src/i18n/en.json (hunk 1 above)
Comment (RitikPandey1) [reply ui/src/i18n/en.json:129]: I have updated the message

**Comment by RitikPandey1** [comment]
> This looks great. Besides the small nitpick below, I think we can improve the "Skip" behaviour: Instead of returning to the `AddToPlaylist` dialog when clicking the "Skip" button, it should just go ahead and add the non-duplicated songs to the playlist, closing all dialogs, similar to when you press the "Add" button.
> 
> Also, if there were no changes to the Go codebase, there should not be any changes to the `go.sum` file. Can you please rebase and remove it from the commit?

All the mentioned changes are implemented but I'm not able to remove the go.sum file from my previous commit, please tell me  how can I remove go.sum file ?

**Comment by deluan** [comment]
Take a look at `git rebase -i` (rebase interactively). If you can't do it, let me know and I can fix this, but this is definitely something you should learn, very helpful.

File: ui/src/i18n/en.json [hunk 2]
```diff
@@ -1,341 +1,346 @@
...
+        "sync": "Auto-import",
+        "path": "Import from"
+      },
+      "actions": {
+        "selectPlaylist": "Select a playlist:",
+        "addNewPlaylist": "Create \"%{name}\"",
+        "export": "Export"
+      },
+      "message": {
+        "duplicate_song": "Add duplicate song ",
```
Comment (deluan) [inline ui/src/i18n/en.json:128]: This should be `"Add duplicated songs`

Comment (RitikPandey1) [reply ui/src/i18n/en.json]: I have changed the message

**Comment by RitikPandey1** [comment]
I just noticed that build was not succesful and showing error ` missing go.sum entry`, Is this because I removed go.sum file from my previous commit? I'm little confuse beacuse I didn't touch go.sum file when I was making initial changes and when I removed it then build wast not succesful, could you please tell me what I did wrong and how can I resolve this issue ? @deluan

**Comment by jvoisin** [comment]
You have to keep the go.sum and go.mod files, without modifying them. maybe you removed them from your filesystem while removing them from your commit?

**Comment by RitikPandey1** [comment]
> You have to keep the go.sum and go.mod files, without modifying them. maybe you removed them from your filesystem while removing them from your commit?

No these files are in my filesystem

**Comment by RitikPandey1** [comment]
I have added unmodified go.sum file

**Comment by RitikPandey1** [comment]
Requesting review
@deluan

**Comment by deluan** [comment]
Found a bug. I had a playlist with one song. If I tried to add 3 songs (including the one that's already there) and press Skip, it worked as expected, only adding the other two. But if I try to add again the same 3 songs (now all of them are in the playlist), and press Skip again, it adds all three again!
Screen Recording 2021-03-17 at 12 24 37 AM

**Comment by RitikPandey1** [comment]
I'll try to debug this.

**Comment by RitikPandey1** [comment]
Bug fixed.
navidromeskipsong

**Comment by RitikPandey1** [comment]
Requesting review 
@deluan

**Comment by deluan** [comment]
Thanks for the PR!

**Pull request merged**